configure.py

```diff
@@ -2982,6 +2998,37 @@ def canonicalize_options(options, info_os, info_arch):
     if options.cpu != options.arch:
         logging.info('Canonicalized CPU target %s to %s', options.cpu, options.arch)
 
+    # select and sanity check build targets
+    def canonicalize_build_targets(options):
+        build_targets = read_build_targets(options)
+
+        # --build-targets was not provided: build default targets
+        if build_targets == None:
+            options.build_targets = "cli,tests"
+            return
+
+        # --build-targets was set to empty string: invalid input
+        if not build_targets:
+            raise UserError("select at least one --build-targets")
+
+        for build_target in build_targets:
+            if build_target not in __acceptable_build_targets:
+                raise UserError("unknown build target: %s" % build_target)
+
+        # building the shared lib desired and without contradiction?
+        if options.build_shared_lib == None:
+            options.build_shared_lib = "shared" in build_targets
+        elif bool(options.build_shared_lib) != bool("shared" in build_targets):
+            raise UserError("inconsistent usage of --enable/disable-shared-library and --build-targets")
+
+        # building the static lib desired and without contradiction?
+        if options.build_static_lib == None:
+            options.build_static_lib = "static" in build_targets
+        elif bool(options.build_static_lib) != bool("static" in build_targets):
+            raise UserError("inconsistent usage of --enable/disable-static-library and --build-targets")
+
+    canonicalize_build_targets(options)
```

Comment:
This turned out to be fairly complex, because it shall keep the default and legacy behaviour (using `--(enable/disable)-(shared/static)-library`)  and consistently introduce `--build-targets=(static/shared)`.

The behaviour for `static` and `shared` lib builds is now as follows:

```
                 options.build_shared_lib
                 None    True   False
 build_targets     
        static    yes     yes    fail
       !static     no    fail     yes
          None     no     yes      no
```

I.e. if the legacy switches are used, they behave as expected (switching library builds on and off). If only the new `--build-targets=` switch is used, it builds exactly what is listed by it. When using a combination of both legacy and new switches, their decisions are checked for consistency. For example this would fail with a `UserError`:

```
./configure.py --disable-shared-library --build-targets="static,tests"
```

Reply:
Thanks for addressing this. It would be nice to be able to deprecate and remove the older options eventually as --build-targets is much cleaner - but doing so will break everyone's build scripts, so we probably have to support both methods for at least a few releases.

Roughly: 2.12: add --build-targets, 2.13: start printing a deprecation warning when older options are used, 2.xx?/3.0?: remove the options.